## Intro

### What makes a country happier?

What makes a country happy? Or.. An easier question. What makes you, as an individual, happier? Your future expectations or your income or your honest government? According to <a href="https://worldhappiness.report/faq/" target="_blank">happiness report</a>.; "**The variables used reflect what has been broadly found in the research literature to be important in explaining national-level differences in life evaluations.**". We will be using those datas to understand better what makes a country more happy. Without further ado, let's get started

***
- I will be using two datasets in this analysis. One of them is latest world happines report of 2021 and the other one contains data from previous years.
***

***
- Let's start with importing required libraries.
***

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt 
import seaborn as sns 
import matplotlib as mpl



import plotly 
import plotly.express as px
import plotly.graph_objs as go
import plotly.offline as py
from plotly.offline import iplot
from plotly.subplots import make_subplots
import plotly.figure_factory as ff

***
- Let's read our data.
***

In [2]:
df_2021 = pd.read_csv("world-happiness-report-2021.csv")
df_history = pd.read_csv("world-happiness-report.csv")
pd.set_option('display.max_columns', None)

***
- First, inspect two dataframes one by one.
***

***
- Let's check statistical information of 2021 data.
***

In [3]:
df_2021.describe()

Ladder score  Standard error of ladder score  upperwhisker  \
count    149.000000                      149.000000    149.000000   
mean       5.532839                        0.058752      5.648007   
std        1.073924                        0.022001      1.054330   
min        2.523000                        0.026000      2.596000   
25%        4.852000                        0.043000      4.991000   
50%        5.534000                        0.054000      5.625000   
75%        6.255000                        0.070000      6.344000   
max        7.842000                        0.173000      7.904000   

       lowerwhisker  Logged GDP per capita  Social support  \
count    149.000000             149.000000      149.000000   
mean       5.417631               9.432208        0.814745   
std        1.094879               1.158601        0.114889   
min        2.449000               6.635000        0.463000   
25%        4.706000               8.541000        0.750000   
50%        5.413000               9.569000        0.832000   
75%        6.128000              10.421000        0.905000   
max        7.780000              11.647000        0.983000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               149.000000                    149.000000  149.000000   
mean                 64.992799                      0.791597   -0.015134   
std                   6.762043                      0.113332    0.150657   
min                  48.478000                      0.382000   -0.288000   
25%                  59.802000                      0.718000   -0.126000   
50%                  66.603000                      0.804000   -0.036000   
75%                  69.600000                      0.877000    0.079000   
max                  76.953000                      0.970000    0.542000   

       Perceptions of corruption  Ladder score in Dystopia  \
count                 149.000000              1.490000e+02   
mean                    0.727450              2.430000e+00   
std                     0.179226              5.347044e-15   
min                     0.082000              2.430000e+00   
25%                     0.667000              2.430000e+00   
50%                     0.781000              2.430000e+00   
75%                     0.845000              2.430000e+00   
max                     0.939000              2.430000e+00   

       Explained by: Log GDP per capita  Explained by: Social support  \
count                        149.000000                    149.000000   
mean                           0.977161                      0.793315   
std                            0.404740                      0.258871   
min                            0.000000                      0.000000   
25%                            0.666000                      0.647000   
50%                            1.025000                      0.832000   
75%                            1.323000                      0.996000   
max                            1.751000                      1.172000   

       Explained by: Healthy life expectancy  \
count                             149.000000   
mean                                0.520161   
std                                 0.213019   
min                                 0.000000   
25%                                 0.357000   
50%                                 0.571000   
75%                                 0.665000   
max                                 0.897000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  149.000000                149.000000   
mean                                     0.498711                  0.178047   
std                                      0.137888                  0.098270   
min                                      0.000000                  0.000000   
25%                                      0.409000                  0.105000   
50%              

***
- What about historic data?
***

In [4]:
df_history.describe()

year  Life Ladder  Log GDP per capita  Social support  \
count  1949.000000  1949.000000         1913.000000     1936.000000   
mean   2013.216008     5.466705            9.368453        0.812552   
std       4.166828     1.115711            1.154084        0.118482   
min    2005.000000     2.375000            6.635000        0.290000   
25%    2010.000000     4.640000            8.464000        0.749750   
50%    2013.000000     5.386000            9.460000        0.835500   
75%    2017.000000     6.283000           10.353000        0.905000   
max    2020.000000     8.019000           11.648000        0.987000   

       Healthy life expectancy at birth  Freedom to make life choices  \
count                       1894.000000                   1917.000000   
mean                          63.359374                      0.742558   
std                            7.510245                      0.142093   
min                           32.300000                      0.258000   
25%                           58.685000                      0.647000   
50%                           65.200000                      0.763000   
75%                           68.590000                      0.856000   
max                           77.100000                      0.985000   

        Generosity  Perceptions of corruption  Positive affect  \
count  1860.000000                1839.000000      1927.000000   
mean      0.000103                   0.747125         0.710003   
std       0.162215                   0.186789         0.107100   
min      -0.335000                   0.035000         0.322000   
25%      -0.113000                   0.690000         0.625500   
50%      -0.025500                   0.802000         0.722000   
75%       0.091000                   0.872000         0.799000   
max       0.698000                   0.983000         0.944000   

       Negative affect  
count      1933.000000  
mean          0.268544  
std           0.085168  
min           0.083000  
25%           0.206000  
50%           0.258000  
75%           0.320000  
max           0.705000

***
- Now, let's check general information of both dataframes.
***

In [5]:
df_2021.info()
print()
print("-"*60)
print()
df_history.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149 entries, 0 to 148
Data columns (total 20 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country name                                149 non-null    object 
 1   Regional indicator                          149 non-null    object 
 2   Ladder score                                149 non-null    float64
 3   Standard error of ladder score              149 non-null    float64
 4   upperwhisker                                149 non-null    float64
 5   lowerwhisker                                149 non-null    float64
 6   Logged GDP per capita                       149 non-null    float64
 7   Social support                              149 non-null    float64
 8   Healthy life expectancy                     149 non-null    float64
 9   Freedom to make life choices                149 non-null    float64
 10  Generosity    

***
- 2021 data has no missing values. Great! On the other hand historic data has some missing values. The other thing is some of the column names are different on each dataframe. Although there are 11 and 20 variables in dataframes, my interest will be on
    - 'Country name'
    - 'Year'
    - 'Life Ladder'
    - 'Log GDP per capita'
    - 'Social support'
    - 'Healthy life expectancy at birth'
    - 'Freedom to make life choices'
    - 'Generosity'
    - 'Perceptions of corruption'
    - 'Regional indicator'
    
- Let's prepare our data and concatenate two dataframes into one dataframe.
***

In [6]:
df_2021 = df_2021.rename(columns={"Ladder score": "Life Ladder", "Logged GDP per capita": "Log GDP per capita", "Healthy life expectancy":"Healthy life expectancy at birth", })
df = pd.concat([df_history, df_2021], axis=0, join="outer", ignore_index=True)
df = df.drop(columns=['Standard error of ladder score', 'upperwhisker', 'lowerwhisker',
       'Ladder score in Dystopia', 'Explained by: Log GDP per capita',
       'Explained by: Social support', 'Explained by: Healthy life expectancy',
       'Explained by: Freedom to make life choices',
       'Explained by: Generosity', 'Explained by: Perceptions of corruption',
       'Dystopia + residual', "Positive affect", "Negative affect"])

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2098 entries, 0 to 2097
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Country name                      2098 non-null   object 
 1   year                              1949 non-null   float64
 2   Life Ladder                       2098 non-null   float64
 3   Log GDP per capita                2062 non-null   float64
 4   Social support                    2085 non-null   float64
 5   Healthy life expectancy at birth  2043 non-null   float64
 6   Freedom to make life choices      2066 non-null   float64
 7   Generosity                        2009 non-null   float64
 8   Perceptions of corruption         1988 non-null   float64
 9   Regional indicator                149 non-null    object 
dtypes: float64(8), object(2)
memory usage: 164.0+ KB


***
- There seems to be some missing values. Let's check missing values.
***

In [8]:
df.isnull().sum()

Country name                           0
year                                 149
Life Ladder                            0
Log GDP per capita                    36
Social support                        13
Healthy life expectancy at birth      55
Freedom to make life choices          32
Generosity                            89
Perceptions of corruption            110
Regional indicator                  1949
dtype: int64

***
- *`year`* column has missing values but all those values are actually 2021. Because we know historic data had no missing values in *`year`* column.
- *`Regional indicator`* colum has 1949 missing values. That is just like *`year`* column. Historic column had no column as *`Regional indicator`*. We will fill those missing values by using 2021 data.
- I will check the other ones later.
***

***
- Now, let's fill *`Regional indicator`* column by using 2021 data.
***

In [9]:
for i in df["Country name"].unique():
    filt = (df["Country name"] == i)
    regional = df.loc[filt, "Regional indicator"].unique()[-1]
    df.loc[filt, "Regional indicator"] = regional
df

Country name    year  Life Ladder  Log GDP per capita  Social support  \
0     Afghanistan  2008.0        3.724               7.370           0.451   
1     Afghanistan  2009.0        4.402               7.540           0.552   
2     Afghanistan  2010.0        4.758               7.647           0.539   
3     Afghanistan  2011.0        3.832               7.620           0.521   
4     Afghanistan  2012.0        3.783               7.705           0.521   
...           ...     ...          ...                 ...             ...   
2093      Lesotho     NaN        3.512               7.926           0.787   
2094     Botswana     NaN        3.467               9.782           0.784   
2095       Rwanda     NaN        3.415               7.676           0.552   
2096     Zimbabwe     NaN        3.145               7.943           0.750   
2097  Afghanistan     NaN        2.523               7.695           0.463   

      Healthy life expectancy at birth  Freedom to make life choices  \
0                               50.800                         0.718   
1                               51.200                         0.679   
2                               51.600                         0.600   
3                               51.920                         0.496   
4                               52.240                         0.531   
...                                ...                           ...   
2093                            48.700                         0.715   
2094                            59.269                         0.824   
2095                            61.400                         0.897   
2096                            56.201                         0.677   
2097                            52.493                         0.382   

      Generosity  Perceptions of corruption  Regional indicator  
0          0.168                      0.882          South Asia  
1          0.190                      0.850          South Asia  
2          0.121                      0.707          South Asia  
3          0.162                      0.731          South Asia  
4          0.236                      0.776          South Asia  
...          ...                        ...                 ...  
2093      -0.131                      0.915  Sub-Saharan Africa  
2094      -0.246                      0.801  Sub-Saharan Africa  
2095       0.061                      0.167  Sub-Saharan Africa  
2096      -0.047                      0.821  Sub-Saharan Africa  
2097      -0.102                      0.924          South Asia  

[2098 rows x 10 columns]

***
- Now it is time for *`year`* column.
***

In [10]:
df.year = df.year.fillna(2021)

***
- Let's check missing values again.
***

In [11]:
df.isnull().sum()

Country name                          0
year                                  0
Life Ladder                           0
Log GDP per capita                   36
Social support                       13
Healthy life expectancy at birth     55
Freedom to make life choices         32
Generosity                           89
Perceptions of corruption           110
Regional indicator                   63
dtype: int64

***
- Still there are some missing values. What should we do? Filling those values with mean or median values may be one solution. Or we can just drop those values which is not the best option considering there are at least 110 rows with missing values.
- I will be filling missing values with that country's mean values. That seems like a proper way to me.
***

In [12]:
def fill_by_country(df):
    missing = df.drop(["Country name", "Regional indicator"], axis=1).isnull().sum()
    missing = missing[missing>0]
    for i in missing.index:
        df[i] = df.groupby("Country name")[i].transform(lambda val: val.fillna(val.mean()))
    return df

In [13]:
fill_by_country(df)

Country name    year  Life Ladder  Log GDP per capita  Social support  \
0     Afghanistan  2008.0        3.724               7.370           0.451   
1     Afghanistan  2009.0        4.402               7.540           0.552   
2     Afghanistan  2010.0        4.758               7.647           0.539   
3     Afghanistan  2011.0        3.832               7.620           0.521   
4     Afghanistan  2012.0        3.783               7.705           0.521   
...           ...     ...          ...                 ...             ...   
2093      Lesotho  2021.0        3.512               7.926           0.787   
2094     Botswana  2021.0        3.467               9.782           0.784   
2095       Rwanda  2021.0        3.415               7.676           0.552   
2096     Zimbabwe  2021.0        3.145               7.943           0.750   
2097  Afghanistan  2021.0        2.523               7.695           0.463   

      Healthy life expectancy at birth  Freedom to make life choices  \
0                               50.800                         0.718   
1                               51.200                         0.679   
2                               51.600                         0.600   
3                               51.920                         0.496   
4                               52.240                         0.531   
...                                ...                           ...   
2093                            48.700                         0.715   
2094                            59.269                         0.824   
2095                            61.400                         0.897   
2096                            56.201                         0.677   
2097                            52.493                         0.382   

      Generosity  Perceptions of corruption  Regional indicator  
0          0.168                      0.882          South Asia  
1          0.190                      0.850          South Asia  
2          0.121                      0.707          South Asia  
3          0.162                      0.731          South Asia  
4          0.236                      0.776          South Asia  
...          ...                        ...                 ...  
2093      -0.131                      0.915  Sub-Saharan Africa  
2094      -0.246                      0.801  Sub-Saharan Africa  
2095       0.061                      0.167  Sub-Saharan Africa  
2096      -0.047                      0.821  Sub-Saharan Africa  
2097      -0.102                      0.924          South Asia  

[2098 rows x 10 columns]

In [14]:
df.isnull().sum()

Country name                         0
year                                 0
Life Ladder                          0
Log GDP per capita                  12
Social support                       1
Healthy life expectancy at birth     4
Freedom to make life choices         0
Generosity                          12
Perceptions of corruption            2
Regional indicator                  63
dtype: int64

***
- We filled most of the missing values but still there are some. Let's check those missing values if the same row has more than one missing values. If that is the case, it may be clever to just drop those rows.
***

In [15]:
df[df["Log GDP per capita"].isnull()]

Country name    year  Life Ladder  Log GDP per capita  \
423                Cuba  2006.0        5.418                 NaN   
1559            Somalia  2014.0        5.528                 NaN   
1560            Somalia  2015.0        5.354                 NaN   
1561            Somalia  2016.0        4.668                 NaN   
1562  Somaliland region  2009.0        4.991                 NaN   
1563  Somaliland region  2010.0        4.657                 NaN   
1564  Somaliland region  2011.0        4.931                 NaN   
1565  Somaliland region  2012.0        5.057                 NaN   
1596        South Sudan  2014.0        3.832                 NaN   
1597        South Sudan  2015.0        4.071                 NaN   
1598        South Sudan  2016.0        2.888                 NaN   
1599        South Sudan  2017.0        2.817                 NaN   

      Social support  Healthy life expectancy at birth  \
423            0.970                             68.44   
1559           0.611                             49.60   
1560           0.599                             50.10   
1561           0.594                             50.00   
1562           0.880                               NaN   
1563           0.829                               NaN   
1564           0.788                               NaN   
1565           0.786                               NaN   
1596           0.545                             49.84   
1597           0.585                             50.20   
1598           0.532                             50.60   
1599           0.557                             51.00   

      Freedom to make life choices  Generosity  Perceptions of corruption  \
423                          0.281         NaN                        NaN   
1559                         0.874         NaN                      0.456   
1560                         0.968         NaN                      0.410   
1561                         0.917         NaN                      0.441   
1562                         0.746         NaN                      0.513   
1563                         0.820         NaN                      0.471   
1564                         0.858         NaN                      0.357   
1565                         0.758         NaN                      0.334   
1596                         0.567         NaN                      0.742   
1597                         0.512         NaN                      0.710   
1598                         0.440         NaN                      0.785   
1599                         0.456         NaN                      0.761   

     Regional indicator  
423                 NaN  
1559                NaN  
1560                NaN  
1561                NaN  
1562                NaN  
1563                NaN  
1564                NaN  
1565                NaN  
1596                NaN  
1597                NaN  
1598                NaN  
1599                NaN

***
- All *`Log GDP per capita`*, *`Generosity`*  and *`Regional indicator`* columns' missing values in the same 12 rows. Also 4 of them also contains missing values in *`Healthy life expectancy at birth`* column. Let' drop those 12 rows.
***

In [16]:
df = df[~df["Log GDP per capita"].isnull()]
df.isnull().sum()

Country name                         0
year                                 0
Life Ladder                          0
Log GDP per capita                   0
Social support                       1
Healthy life expectancy at birth     0
Freedom to make life choices         0
Generosity                           0
Perceptions of corruption            1
Regional indicator                  51
dtype: int64

***
- Apart from *`Regional indicator`* column, we have only 2 missing values. Let's look one of them.
***

In [17]:
df[df["Social support"].isnull()]

Country name    year  Life Ladder  Log GDP per capita  Social support  \
1310         Oman  2011.0        6.853              10.382             NaN   

      Healthy life expectancy at birth  Freedom to make life choices  \
1310                              65.5                         0.916   

      Generosity  Perceptions of corruption Regional indicator  
1310       0.025                        NaN                NaN

***
- Great! Our last missing values in the *`Social support`* and *`Perceptions of corruption`* in the same row. Let's drop this row too.
***

In [18]:
df = df[~df["Social support"].isnull()]
df.isnull().sum()

Country name                         0
year                                 0
Life Ladder                          0
Log GDP per capita                   0
Social support                       0
Healthy life expectancy at birth     0
Freedom to make life choices         0
Generosity                           0
Perceptions of corruption            0
Regional indicator                  50
dtype: int64

***
- We have done it! There is no missing values in the columns other than *`Regional indicator`* column in which it is OK to leave as NaN. This is because 2021 data had no information about those countries that historic data had. It is not a problem though. We can start our analysis now.
***

***
## What Makes a Country Happier?
***

def trust(corrupt):
    if corrupt >=  0.8450:
        return "Low trust in institutions"
    elif corrupt < 0.8450 and corrupt > 0.781:
        return "Lower than normal level trust in institutions"
    elif corrupt <= 0.781 and corrupt > 0.667:
        return "About the normal level trust in institutions"
    else:
        return "High level trust in institutions"
df3["Trust in institutions"] = df3["Perceptions of corruption"].apply(trust)

***
- First, let's check correlation among the numerical variables in the dataset.
***

In [19]:
df.drop("year",axis=1).corr()

Life Ladder  Log GDP per capita  \
Life Ladder                          1.000000            0.788567   
Log GDP per capita                   0.788567            1.000000   
Social support                       0.711121            0.695889   
Healthy life expectancy at birth     0.737231            0.849697   
Freedom to make life choices         0.525287            0.366862   
Generosity                           0.178586           -0.007500   
Perceptions of corruption           -0.445101           -0.372409   

                                  Social support  \
Life Ladder                             0.711121   
Log GDP per capita                      0.695889   
Social support                          1.000000   
Healthy life expectancy at birth        0.612149   
Freedom to make life choices            0.416092   
Generosity                              0.056169   
Perceptions of corruption              -0.228155   

                                  Healthy life expectancy at birth  \
Life Ladder                                               0.737231   
Log GDP per capita                                        0.849697   
Social support                                            0.612149   
Healthy life expectancy at birth                          1.000000   
Freedom to make life choices                              0.395658   
Generosity                                                0.024210   
Perceptions of corruption                                -0.348888   

                                  Freedom to make life choices  Generosity  \
Life Ladder                                           0.525287    0.178586   
Log GDP per capita                                    0.366862   -0.007500   
Social support                                        0.416092    0.056169   
Healthy life expectancy at birth                      0.395658    0.024210   
Freedom to make life choices                          1.000000    0.315045   
Generosity                                            0.315045    1.000000   
Perceptions of corruption                            -0.484465   -0.281922   

                                  Perceptions of corruption  
Life Ladder                                       -0.445101  
Log GDP per capita                                -0.372409  
Social support                                    -0.228155  
Healthy life expectancy at birth                  -0.348888  
Freedom to make life choices                      -0.484465  
Generosity                                        -0.281922  
Perceptions of corruption                          1.000000

***
- Happiness score(Life Ladder) has strong level correlation with GDP, Social Support and Healthy life expectancy at birth.
- Freedom to make life choices and happiness score have mide level correlation between them.
- Perception of corruption and happiness score have weak level negative level correlation between them.
***

***
- Let's see correlation in the heatmap.
***

In [40]:
fig = go.Figure(go.Heatmap(z=df.corr(), x=df.corr().columns.tolist(), y=df.corr().columns.tolist(),
                          colorscale="viridis"))
fig.show()

***
- What about happines at the Regional level?
***

In [27]:
df.groupby("Regional indicator")["Life Ladder"].describe().sort_values(by="std")

count      mean       std    min      25%  \
Regional indicator                                                              
North America and ANZ                62.0  7.254919  0.179134  6.804  7.16675   
East Asia                            88.0  5.645511  0.563235  4.454  5.28150   
Commonwealth of Independent States  182.0  5.212835  0.612177  3.675  4.74650   
Central and Eastern Europe          242.0  5.557930  0.619974  3.844  5.10925   
Sub-Saharan Africa                  426.0  4.317688  0.633168  2.694  3.83100   
South Asia                           91.0  4.537363  0.684750  2.375  4.22900   
Western Europe                      292.0  6.833096  0.705528  4.720  6.43800   
Southeast Asia                      125.0  5.345960  0.752216  3.569  4.88000   
Latin America and Caribbean         299.0  5.987666  0.766028  3.352  5.64950   
Middle East and North Africa        228.0  5.384469  0.980232  2.983  4.70100   

                                       50%      75%    max  
Regional indicator                                          
North America and ANZ               7.2555  7.37825  7.650  
East Asia                           5.7220  5.97900  6.947  
Commonwealth of Independent States  5.2565  5.71050  6.568  
Central and Eastern Europe          5.5915  6.03575  7.034  
Sub-Saharan Africa                  4.3200  4.77950  6.241  
South Asia                          4.5670  5.02600  5.831  
Western Europe                      6.9755  7.41700  8.019  
Southeast Asia                      5.3390  5.88500  7.062  
Latin America and Caribbean         6.0250  6.46650  7.615  
Middle East and North Africa        5.1650  6.17475  7.433

***
- North America and ANZ	has the highest level happines mean on the other hand Sub-Saharan Africa has the least level happines mean.
- Western Europe with the highest level happiness and South Asia least highest levet happiness.
- North America and ANZ has the least standard deviation and almost the same mediand and mean which means happiness there is a normal distribution acroos countires in North America and ANZ.
- Middle East and North Africa with the highest standard deviation.
***

***
- Let's see all this better with boxplot.
***

In [39]:
fig = px.box(df, x="Life Ladder", y="Regional indicator", hover_data = df[['Regional indicator','Country name']])
fig.show()

***
- South Asia, Latin America and Caribbean, North American and ANZ, Western Europe has several outliers in the minimum side.
- Sub-Saharan Africa has outlier in the maxiumum side.
- Middle East and North Africa seems interesting to me since it has long tail in both minimum and maximum side.
- Let's dive into Middle East and North Africa.

In [36]:
middle_east = df[df["Regional indicator"] == "Middle East and North Africa"]
middle_east

Country name    year  Life Ladder  Log GDP per capita  \
25                    Algeria  2010.0        5.464               9.287   
26                    Algeria  2011.0        5.317               9.297   
27                    Algeria  2012.0        5.605               9.311   
28                    Algeria  2014.0        6.355               9.335   
29                    Algeria  2016.0        5.341               9.362   
...                       ...     ...          ...                 ...   
2071                  Lebanon  2021.0        4.584               9.626   
2073  Palestinian Territories  2021.0        4.517               8.485   
2075                   Jordan  2021.0        4.395               9.182   
2080                    Egypt  2021.0        4.283               9.367   
2089                    Yemen  2021.0        3.658               7.578   

      Social support  Healthy life expectancy at birth  \
25          0.803375                            64.500   
26          0.810000                            64.660   
27          0.839000                            64.820   
28          0.818000                            65.140   
29          0.749000                            65.500   
...              ...                               ...   
2071        0.848000                            67.355   
2073        0.826000                            62.250   
2075        0.767000                            67.000   
2080        0.750000                            61.998   
2089        0.832000                            57.122   

      Freedom to make life choices  Generosity  Perceptions of corruption  \
25                        0.593000   -0.205000                   0.618000   
26                        0.530000   -0.181000                   0.638000   
27                        0.587000   -0.172000                   0.690000   
28                        0.513571   -0.133286                   0.699714   
29                        0.513571   -0.133286                   0.699714   
...                            ...         ...                        ...   
2071                      0.525000   -0.073000                   0.898000   
2073                      0.653000   -0.163000                   0.821000   
2075                      0.755000   -0.167000                   0.705000   
2080                      0.749000   -0.182000                   0.795000   
2089                      0.602000   -0.147000                   0.800000   

                Regional indicator  
25    Middle East and North Africa  
26    Middle East and North Africa  
27    Middle East and North Africa  
28    Middle East and North Africa  
29    Middle East and North Africa  
...                            ...  
2071  Middle East and North Africa  
2073  Middle East and North Africa  
2075  Middle East and North Africa  
2080  Middle East and North Africa  
2089  Middle East and North Africa  

[228 rows x 10 columns]

***
- Let's see how correlated the variables in Middle East and North Africa.
***

In [37]:
df.drop("year",axis=1).corr()

Life Ladder  Log GDP per capita  \
Life Ladder                          1.000000            0.788567   
Log GDP per capita                   0.788567            1.000000   
Social support                       0.711121            0.695889   
Healthy life expectancy at birth     0.737231            0.849697   
Freedom to make life choices         0.525287            0.366862   
Generosity                           0.178586           -0.007500   
Perceptions of corruption           -0.445101           -0.372409   

                                  Social support  \
Life Ladder                             0.711121   
Log GDP per capita                      0.695889   
Social support                          1.000000   
Healthy life expectancy at birth        0.612149   
Freedom to make life choices            0.416092   
Generosity                              0.056169   
Perceptions of corruption              -0.228155   

                                  Healthy life expectancy at birth  \
Life Ladder                                               0.737231   
Log GDP per capita                                        0.849697   
Social support                                            0.612149   
Healthy life expectancy at birth                          1.000000   
Freedom to make life choices                              0.395658   
Generosity                                                0.024210   
Perceptions of corruption                                -0.348888   

                                  Freedom to make life choices  Generosity  \
Life Ladder                                           0.525287    0.178586   
Log GDP per capita                                    0.366862   -0.007500   
Social support                                        0.416092    0.056169   
Healthy life expectancy at birth                      0.395658    0.024210   
Freedom to make life choices                          1.000000    0.315045   
Generosity                                            0.315045    1.000000   
Perceptions of corruption                            -0.484465   -0.281922   

                                  Perceptions of corruption  
Life Ladder                                       -0.445101  
Log GDP per capita                                -0.372409  
Social support                                    -0.228155  
Healthy life expectancy at birth                  -0.348888  
Freedom to make life choices                      -0.484465  
Generosity                                        -0.281922  
Perceptions of corruption                          1.000000

In [38]:
middle_east.drop("year",axis=1).corr()

Life Ladder  Log GDP per capita  \
Life Ladder                          1.000000            0.818969   
Log GDP per capita                   0.818969            1.000000   
Social support                       0.640044            0.559865   
Healthy life expectancy at birth     0.674217            0.655847   
Freedom to make life choices         0.524962            0.544174   
Generosity                           0.448896            0.453228   
Perceptions of corruption           -0.526460           -0.630186   

                                  Social support  \
Life Ladder                             0.640044   
Log GDP per capita                      0.559865   
Social support                          1.000000   
Healthy life expectancy at birth        0.372988   
Freedom to make life choices            0.313726   
Generosity                              0.237148   
Perceptions of corruption              -0.407732   

                                  Healthy life expectancy at birth  \
Life Ladder                                               0.674217   
Log GDP per capita                                        0.655847   
Social support                                            0.372988   
Healthy life expectancy at birth                          1.000000   
Freedom to make life choices                              0.387367   
Generosity                                                0.382277   
Perceptions of corruption                                -0.196717   

                                  Freedom to make life choices  Generosity  \
Life Ladder                                           0.524962    0.448896   
Log GDP per capita                                    0.544174    0.453228   
Social support                                        0.313726    0.237148   
Healthy life expectancy at birth                      0.387367    0.382277   
Freedom to make life choices                          1.000000    0.349958   
Generosity                                            0.349958    1.000000   
Perceptions of corruption                            -0.565860   -0.234272   

                                  Perceptions of corruption  
Life Ladder                                       -0.526460  
Log GDP per capita                                -0.630186  
Social support                                    -0.407732  
Healthy life expectancy at birth                  -0.196717  
Freedom to make life choices                      -0.565860  
Generosity                                        -0.234272  
Perceptions of corruption                          1.000000

***
- Even though some similarities can be found with the whole dataset correlation, in Middle East and North Africa has some differences in correlation matrix.
- GDP has more correlation with happiness compared to whole dataset but in a low margin but on the other hand, Generosity has so much more correlation with happines in Middle East and North Africa compared to whole dataset.
- Social support, Healthy life expectancy at birth and Perception of corruption has much lower correlation with happines than whole dataset.
- Freedom to make life choices' correlation almost identical in both datasets.
***

***
- Let's look at heatmap for better understanding.
***

In [41]:
fig = go.Figure(go.Heatmap(z=middle_east.corr(), x=middle_east.corr().columns.tolist(), y=middle_east.corr().columns.tolist(),
                          colorscale="viridis"))
fig.show()

In [42]:
middle_east.describe()

year  Life Ladder  Log GDP per capita  Social support  \
count   228.000000   228.000000          228.000000      228.000000   
mean   2014.000000     5.384469            9.692509        0.790107   
std       4.412637     0.980232            0.872271        0.092880   
min    2005.000000     2.983000            7.578000        0.535000   
25%    2011.000000     4.701000            9.219750        0.726500   
50%    2014.000000     5.165000            9.490000        0.818000   
75%    2018.000000     6.174750           10.571250        0.862250   
max    2021.000000     7.433000           11.367000        0.959000   

       Healthy life expectancy at birth  Freedom to make life choices  \
count                        228.000000                    228.000000   
mean                          64.557434                      0.681605   
std                            3.889171                      0.139725   
min                           53.400000                      0.315000   
25%                           62.287500                      0.594500   
50%                           65.200000                      0.664500   
75%                           66.500000                      0.774000   
max                           73.700000                      0.962000   

       Generosity  Perceptions of corruption  
count  228.000000                 228.000000  
mean    -0.058448                   0.724934  
std      0.123149                   0.149805  
min     -0.244000                   0.203000  
25%     -0.151000                   0.666500  
50%     -0.081000                   0.753500  
75%      0.031000                   0.831250  
max      0.241000                   0.949000

***
- Based on descriptive information, possible outliers can be seen in the:
    - Healthy life expectancy at birth
    - Freedom to make life choices
    - Perceptions of corruption
***

***
- In this EDA, I will mostly focus on happines score.
***

## Happiness in the Middle East and North Africa

***
- Let's start with boxplot.
***

In [48]:
fig = px.box(middle_east, x="Life Ladder", hover_data = middle_east[['Country name']])
fig.update_traces(quartilemethod="inclusive")
fig.show()

***
- Boxplot does not show any possible outliers.
***

***
- Let's continue with a bar plot.
***

In [53]:
middle_east = middle_east.sort_values(by="Life Ladder")

fig = px.bar(middle_east, x="Life Ladder", y="Country name")
fig.show()

***
- **Israel** has the highest happiness score in Middle East and North Africa.
- **Yemen** has the least happiness score in Middle East and North Africa.

In [62]:
fig=go.Figure()
fig.add_trace(go.Scatter(
    x=middle_east.groupby("Country name").mean().sort_values(by="Life Ladder").index,
    y=middle_east.groupby("Country name").mean().sort_values(by="Life Ladder")["Life Ladder"],
    name='Happines Score',
    mode='markers+text',
    marker_color='blue',
    marker_size=10,
    textposition='top center',
    line=dict(color='red',dash='dash'),
))
fig.update_layout(
    title= "<b>Middle East Happiness Score in 2021</b>",
    xaxis_title="<b>Country</b>",
    yaxis_title="<b>Happiness Score</b>",
    template='plotly_white',
    font=dict(
        size=12,
        color="Black",
        family="Oswald', sans-serif"
        ),
    xaxis=dict(showgrid=True),
    yaxis=dict(showgrid=True),
    yaxis2=dict(showgrid=True,overlaying='y',side='right',title='<b>Happiness Score</b>'),
)
fig.show()

***
- It is important to understand that we are not working with just 2021 data but all data which contains data from 2013 to 2021. This values are mean of all the years.
- Most of the countries has happines score between 4.5 and 6.5.
***

***
- Let's see happiness score trends in the past years.
***

In [99]:
middle_east = middle_east.sort_values(["Country name", "year"])

In [75]:
css3_colors = ['#add8e6', '#f08080','#e0ffff','#fafad2','#d3d3d3','#90ee90','#ffb6c1','#ffa07a','#20b2aa','#87cefa','#778899','#b0c4de','#32cd32','#ff00ff','#66cdaa','#ba55d3', '#7b68ee']
css3_dict ={}
i=0
for name in middle_east["Country name"].unique():
    css3_dict[name]=css3_colors[i]
    i+=1

In [82]:
fig=go.Figure()
for name in middle_east['Country name'].unique():
    fig.add_trace(go.Scatter(
    x=middle_east[middle_east['Country name']==name]['year'],
    y=middle_east[middle_east['Country name']==name]['Life Ladder'],
    name=name,
    mode='markers+text+lines',
    marker_color='black',
    line=dict(color=css3_dict[name]),
    marker_size=3,
    yaxis='y1'))
    
fig.update_layout(
    title="Happiness Score Trend in Central and Eastern Europe ",
    xaxis_title="Year",
    yaxis_title='Happiness Score',
    template='plotly_white',
    font=dict(
        size=14,
        color="Blue",
        family="Oswald', sans-serif"
    ),
    xaxis=dict(showgrid=True),
    yaxis=dict(showgrid=True)
)
fig.show()

***
- This plot seems kind of complicated. Let's visualize the change by country.
***

***
- This visualizations seems better.
- Algeria, Jordan, Turkey and United Arab Emirates have downtrend. 
- Bahrain and Iraq's happiness increased over the last years.
***

***
- Let's start working with other variables' relation with happiness score.
***

## GDP's Impact on Happiness Score

***
- Is really the money answer to our happiness? Let's check!
***

In [87]:
middle_east[middle_east["year"]==2021]

Country name    year  Life Ladder  Log GDP per capita  \
2057                  Algeria  2021.0        4.887               9.342   
1970                  Bahrain  2021.0        6.647              10.669   
2080                    Egypt  2021.0        4.283               9.367   
2066                     Iran  2021.0        4.721               9.584   
2059                     Iraq  2021.0        4.854               9.240   
1960                   Israel  2021.0        7.157              10.575   
2075                   Jordan  2021.0        4.395               9.182   
1995                   Kuwait  2021.0        6.106              10.817   
2071                  Lebanon  2021.0        4.584               9.626   
2028                    Libya  2021.0        5.410               9.622   
2054                  Morocco  2021.0        4.918               8.903   
2073  Palestinian Territories  2021.0        4.517               8.485   
1974             Saudi Arabia  2021.0        6.494              10.743   
2070                  Tunisia  2021.0        4.596               9.266   
2052                   Turkey  2021.0        4.948              10.240   
1973     United Arab Emirates  2021.0        6.561              11.085   
2089                    Yemen  2021.0        3.658               7.578   

      Social support  Healthy life expectancy at birth  \
2057           0.802                            66.005   
1970           0.862                            69.495   
2080           0.750                            61.998   
2066           0.710                            66.300   
2059           0.746                            60.583   
1960           0.939                            73.503   
2075           0.767                            67.000   
1995           0.843                            66.900   
2071           0.848                            67.355   
2028           0.827                            62.300   
2054           0.560                            66.208   
2073           0.826                            62.250   
1974           0.891                            66.603   
2070           0.691                            67.201   
2052           0.822                            67.199   
1973           0.844                            67.333   
2089           0.832                            57.122   

      Freedom to make life choices  Generosity  Perceptions of corruption  \
2057                         0.480      -0.067                      0.752   
1970                         0.925       0.089                      0.722   
2080                         0.749      -0.182                      0.795   
2066                         0.608       0.218                      0.714   
2059                         0.630      -0.053                      0.875   
1960                         0.800       0.031                      0.753   
2075                         0.755      -0.167                      0.705   
1995                         0.867      -0.104                      0.736   
2071                         0.525      -0.073                      0.898   
2028                         0.771      -0.087                      0.667   
2054                         0.774      -0.236                      0.801   
2073                         0.653      -0.163                      0.821   
1974                         0.877      -0.149                      0.684   
2070                         0.656      -0.201                      0.870   
2052                         0.576      -0.139                      0.776   
1973                         0.932       0.074                      0.589   
2089                         0.602      -0.147                      0.800   

                Regional indicator  
2057  Middle East and North Africa  
1970  Middle East and North Africa  
2080  Middle East and North Africa  
2066  Middle East and North Africa  
2059  Middle East and North Africa  
1960  Middle East and North Af

In [88]:
trace = go.Scatter(x=middle_east[middle_east["year"]==2021]['Life Ladder'],y=middle_east[middle_east["year"]==2021]['Log GDP per capita'],text = middle_east['Country name'],mode='markers',marker={'color':'blue', 'size':10})
df=[trace]
layout = go.Layout(title='Happiness Score & Logged GDP per capita in Middle East',xaxis=dict(title='Ladder Score'),yaxis=dict(title='Log GDP per capita'),hovermode='closest')
figure = go.Figure(data=df,layout=layout)
figure.update_layout(template='plotly_white',
                  font=dict(family="Oswald', sans-serif"))
figure.show()

In [89]:
print(f"Happines score and GDP has {round(middle_east.corr().loc['Life Ladder', 'Log GDP per capita'],2)} correlation score.")

Happines score and GDP has 0.82 correlation score.


***
- The answer the question I asked before is yes. GDP has really strong correlation with happines score. We can see this from scatterplot.

***
- What about social support?
***

## Social Support's Impact on Happiness Score

In [90]:
trace = go.Scatter(x=middle_east[middle_east["year"]==2021]['Life Ladder'],y=middle_east[middle_east["year"]==2021]['Social support'],text = middle_east['Country name'],mode='markers',marker={'color':'blue', 'size':10})
df=[trace]
layout = go.Layout(title='Happiness Score & Social support in Middle East',xaxis=dict(title='Ladder Score'),yaxis=dict(title='Social support'),hovermode='closest')
figure = go.Figure(data=df,layout=layout)
figure.update_layout(template='plotly_white',
                  font=dict(family="Oswald', sans-serif"))
figure.show()

In [91]:
print(f"Happines score and social support has {round(middle_east.corr().loc['Life Ladder', 'Social support'],2)} correlation score.")

Happines score and social support has 0.64 correlation score.


***
- Also social support has strong correlation score with happiness.
***

## Healthy Life Expectancy at Birth's Impact on Happiness Score

In [92]:
trace = go.Scatter(x=middle_east[middle_east["year"]==2021]['Life Ladder'],y=middle_east[middle_east["year"]==2021]['Healthy life expectancy at birth'],text = middle_east['Country name'],mode='markers',marker={'color':'blue', 'size':10})
df=[trace]
layout = go.Layout(title='Happiness Score & Healthy life expectancy at birth in Middle East',xaxis=dict(title='Ladder Score'),yaxis=dict(title='Healthy life expectancy at birth'),hovermode='closest')
figure = go.Figure(data=df,layout=layout)
figure.update_layout(template='plotly_white',
                  font=dict(family="Oswald', sans-serif"))
figure.show()

In [93]:
print(f"Happines score and healthy life expectancy at birth has {round(middle_east.corr().loc['Life Ladder', 'Healthy life expectancy at birth'],2)} correlation score.")

Happines score and healthy life expectancy at birth has 0.67 correlation score.


***
- Healthy life expectancy at birth has strong correlation score with happiness.
***

## Perceptions of Corruption's Impact on Happiness Score

In [94]:
trace = go.Scatter(x=middle_east[middle_east["year"]==2021]['Life Ladder'],y=middle_east[middle_east["year"]==2021]['Perceptions of corruption'],text = middle_east['Country name'],mode='markers',marker={'color':'blue', 'size':10})
df=[trace]
layout = go.Layout(title='Happiness Score & Perceptions of corruption in Middle East',xaxis=dict(title='Ladder Score'),yaxis=dict(title='Perceptions of corruption'),hovermode='closest')
figure = go.Figure(data=df,layout=layout)
figure.update_layout(template='plotly_white',
                  font=dict(family="Oswald', sans-serif"))
figure.show()

In [95]:
print(f"Happines score and perceptions of corruption has {round(middle_east.corr().loc['Life Ladder', 'Perceptions of corruption'],2)} correlation score.")

Happines score and perceptions of corruption has -0.53 correlation score.


***
- As expected, perceptions of corruption has strong negative correlation with happiness. If people who are ruling you are not honest, it is hard to stay happy.
***

## Freedom to Make Life Choices' Impact on Happiness Score

In [96]:
trace = go.Scatter(x=middle_east[middle_east["year"]==2021]['Life Ladder'],y=middle_east[middle_east["year"]==2021]['Freedom to make life choices'],text = middle_east['Country name'],mode='markers',marker={'color':'blue', 'size':10})
df=[trace]
layout = go.Layout(title='Happiness Score & Freedom to make life choices in Middle East',xaxis=dict(title='Ladder Score'),yaxis=dict(title='Freedom to make life choices'),hovermode='closest')
figure = go.Figure(data=df,layout=layout)
figure.update_layout(template='plotly_white',
                  font=dict(family="Oswald', sans-serif"))
figure.show()

In [97]:
print(f"Happines score and freedom to make life choices has {round(middle_east.corr().loc['Life Ladder', 'Freedom to make life choices'],2)} correlation score.")

Happines score and freedom to make life choices has 0.52 correlation score.


***

We have come to an end of another great analysis. It was really enjoyable for me. It was a quite pleasure to work with this dataset. I would like to thank dataset contibutor for this data. I hope you enjoyed too. If you liked my EDA on this dataset, feel free to check my other notebooks as well. Looking forward for your feedback. Thanks a lot.

Have a great day.